File: frontend/src/features/config/ruleEngine.ts
// Gemelo TS del motor de reglas del backend (app/services/rule_engine.py). MISMA semántica: una
// regla debe evaluarse idéntico en el cliente (reglas de formulario en vivo) y en el backend
// (reglas de validación al guardar). Si cambias uno, cambia el otro y sus tests.

export type ConditionOp =
  | 'igual'
  | 'distinto'
  | 'mayor'
  | 'menor'
  | 'mayor_igual'
  | 'menor_igual'
  | 'en_lista'
  | 'contiene'
  | 'vacio'
  | 'no_vacio'

export interface RuleConditionItem {
  field: string
  op: ConditionOp
  value?: unknown
}

export interface RuleCondition {
  match: 'all' | 'any'
  conditions: RuleConditionItem[]
}

export type RuleContext = Record<string, unknown>

function asNumber(value: unknown): number | null {
  if (value === '' || value === null || value === undefined) return null
  const n = Number(value)
  return Number.isNaN(n) ? null : n
}

function isBlank(value: unknown): boolean {
  return value === null || value === undefined || (typeof value === 'string' && value.trim() === '')
}

function evalSingle(fieldValue: unknown, op: ConditionOp, target: unknown): boolean {
  switch (op) {
    case 'vacio':
      return isBlank(fieldValue)
    case 'no_vacio':
      return !isBlank(fieldValue)
    case 'igual':
      return String(fieldValue) === String(target)
    case 'distinto':
      return String(fieldValue) !== String(target)
    case 'contiene':
      return target != null && String(fieldValue ?? '').includes(String(target))
    case 'en_lista': {
      const options = Array.isArray(target) ? target : String(target).split(',')
      return options.map((o) => String(o).trim()).includes(String(fieldValue))
    }
    default: {
      const a = asNumber(fieldValue)
      const b = asNumber(target)
      if (a === null || b === null) return false
      if (op === 'mayor') return a > b
      if (op === 'menor') return a < b
      if (op === 'mayor_igual') return a >= b
      if (op === 'menor_igual') return a <= b
      return false
    }
  }
}

export function evaluateCondition(condition: RuleCondition | null | undefined, context: RuleContext): boolean {
  if (!condition) return true
  const conditions = condition.conditions ?? []
  if (conditions.length === 0) return true
  const results = conditions.map((c) => evalSingle(context[c.field], c.op, c.value))
  return condition.match === 'any' ? results.some(Boolean) : results.every(Boolean)
}

// --- Evaluador de fórmulas aritméticas seguro (acción "calcular"), sin eval ---

const TOKEN_RE = /\s*(?:(\d+\.?\d*)|([A-Za-z_][A-Za-z0-9_.]*)|([-+*/()]))/y
const PRECEDENCE: Record<string, number> = { '+': 1, '-': 1, '*': 2, '/': 2 }

type Token = { kind: 'num' | 'ref' | 'op'; val: string }

function tokenize(expr: string): Token[] {
  const tokens: Token[] = []
  let pos = 0
  while (pos < expr.length) {
    if (/\s/.test(expr[pos])) {
      pos += 1
      continue
    }
    TOKEN_RE.lastIndex = pos
    const m = TOKEN_RE.exec(expr)
    if (!m || m.index !== pos || m[0].length === 0) return [] // token inválido -> fórmula vacía
    if (m[1] !== undefined) tokens.push({ kind: 'num', val: m[1] })
    else if (m[2] !== undefined) tokens.push({ kind: 'ref', val: m[2] })
    else tokens.push({ kind: 'op', val: m[3] })
    pos = TOKEN_RE.lastIndex
  }
  return tokens
}

/** Evalúa `campoA * 2 + campoB` resolviendo referencias desde `context`. Devuelve null si alguna
 *  referencia no es numérica o la fórmula es inválida (la acción "calcular" no autocompleta). */
export function evaluateFormula(expr: string, context: RuleContext): number | null {
  if (!expr || !expr.trim()) return null
  const tokens = tokenize(expr)
  if (tokens.length === 0) return null

  const output: number[] = []
  const ops: string[] = []

  const applyOp = (): boolean => {
    const op = ops.pop()!
    if (output.length < 2) return false
    const b = output.pop()!
    const a = output.pop()!
    if (op === '+') output.push(a + b)
    else if (op === '-') output.push(a - b)
    else if (op === '*') output.push(a * b)
    else if (op === '/') output.push(b !== 0 ? a / b : 0)
    return true
  }

  for (const t of tokens) {
    if (t.kind === 'num') output.push(Number(t.val))
    else if (t.kind === 'ref') {
      const ref = asNumber(context[t.val])
      if (ref === null) return null
      output.push(ref)
    } else if (t.val === '(') ops.push(t.val)
    else if (t.val === ')') {
      while (ops.length && ops[ops.length - 1] !== '(') if (!applyOp()) return null
      if (!ops.length) return null
      ops.pop()
    } else {
      while (ops.length && PRECEDENCE[ops[ops.length - 1]] >= PRECEDENCE[t.val]) if (!applyOp()) return null
      ops.push(t.val)
    }
  }
  while (ops.length) {
    if (ops[ops.length - 1] === '(' || ops[ops.length - 1] === ')') return null
    if (!applyOp()) return null
  }
  if (output.length !== 1) return null
  return Math.round(output[0] * 1e6) / 1e6
}

File: backend/app/services/rule_engine.py
"""Motor de reglas sin código (Config-B).

Dos funciones puras (sin DB, sin red) — mismo criterio que `custom_fields.py` y las fórmulas de
nómina/fatiga: se testean con datos fijos y tienen una gemela en TS (`frontend/src/features/config/
ruleEngine.ts`) con la MISMA semántica, para que una regla se evalúe idéntico en el cliente (reglas
de formulario en vivo) y en el backend (reglas de validación al guardar).

- `evaluate_condition(condition, context)` -> bool
- `evaluate_formula(expr, context)` -> float | None   (aritmética segura, sin `eval`)
"""

from typing import Any

CONDITION_OPS = (
    "igual",
    "distinto",
    "mayor",
    "menor",
    "mayor_igual",
    "menor_igual",
    "en_lista",
    "contiene",
    "vacio",
    "no_vacio",
)


def _as_number(value: Any) -> float | None:
    try:
        return float(value)
    except (TypeError, ValueError):
        return None


def _is_blank(value: Any) -> bool:
    return value is None or (isinstance(value, str) and value.strip() == "")


def _eval_single(field_value: Any, op: str, target: Any) -> bool:
    if op == "vacio":
        return _is_blank(field_value)
    if op == "no_vacio":
        return not _is_blank(field_value)
    if op == "igual":
        return str(field_value) == str(target)
    if op == "distinto":
        return str(field_value) != str(target)
    if op == "contiene":
        return target is not None and str(target) in str(field_value or "")
    if op == "en_lista":
        options = target if isinstance(target, list) else str(target).split(",")
        return str(field_value) in [str(o).strip() for o in options]
    # operadores numéricos
    a, b = _as_number(field_value), _as_number(target)
    if a is None or b is None:
        return False
    if op == "mayor":
        return a > b
    if op == "menor":
        return a < b
    if op == "mayor_igual":
        return a >= b
    if op == "menor_igual":
        return a <= b
    return False  # pragma: no cover - operador desconocido


def evaluate_condition(condition: dict[str, Any] | None, context: dict[str, Any]) -> bool:
    """Evalúa `{match: "all"|"any", conditions: [{field, op, value}]}` contra `context`
    (dict campo->valor, con los custom aplanados como `custom.<key>`). Una condición vacía o sin
    sub-condiciones se considera SIEMPRE verdadera (la regla aplica incondicionalmente)."""
    if not condition:
        return True
    conditions = condition.get("conditions") or []
    if not conditions:
        return True
    match = condition.get("match", "all")

    results = (
        _eval_single(context.get(c["field"]), c["op"], c.get("value"))
        for c in conditions
    )
    return all(results) if match == "all" else any(results)


# --- Evaluador de fórmulas aritméticas seguro (para la acción "calcular") ---
# Gramática mínima: números, referencias a campos (identificadores, incl. `custom.x`), + - * / y
# paréntesis. Se tokeniza y evalúa con shunting-yard — NUNCA `eval`, así "calcular" no ejecuta
# código arbitrario del usuario.

import re

_TOKEN_RE = re.compile(r"\s*(?:(\d+\.?\d*)|([A-Za-z_][A-Za-z0-9_.]*)|([-+*/()]))")
_PRECEDENCE = {"+": 1, "-": 1, "*": 2, "/": 2}


class FormulaError(ValueError):
    pass


def _tokenize(expr: str) -> list[tuple[str, str]]:
    tokens: list[tuple[str, str]] = []
    pos = 0
    while pos < len(expr):
        if expr[pos].isspace():
            pos += 1
            continue
        m = _TOKEN_RE.match(expr, pos)
        if not m or m.start() == m.end():
            raise FormulaError(f"Token inválido en la fórmula: {expr[pos:]}")
        num, ident, opsym = m.groups()
        if num is not None:
            tokens.append(("num", num))
        elif ident is not None:
            tokens.append(("ref", ident))
        else:
            tokens.append(("op", opsym))
        pos = m.end()
    return tokens


def evaluate_formula(expr: str, context: dict[str, Any]) -> float | None:
    """Evalúa una fórmula aritmética (`campoA * 2 + campoB`) resolviendo referencias desde
    `context`. Devuelve None si alguna referencia no es numérica (campo vacío) — la acción
    "calcular" simplemente no autocompleta en ese caso."""
    if not expr or not expr.strip():
        return None
    tokens = _tokenize(expr)

    output: list[float] = []
    ops: list[str] = []

    def apply_op() -> None:
        op = ops.pop()
        if len(output) < 2:
            raise FormulaError("Fórmula mal formada")
        b = output.pop()
        a = output.pop()
        if op == "+":
            output.append(a + b)
        elif op == "-":
            output.append(a - b)
        elif op == "*":
            output.append(a * b)
        elif op == "/":
            output.append(a / b if b != 0 else 0.0)

    for kind, val in tokens:
        if kind == "num":
            output.append(float(val))
        elif kind == "ref":
            ref = _as_number(context.get(val))
            if ref is None:
                return None
            output.append(ref)
        elif val == "(":
            ops.append(val)
        elif val == ")":
            while ops and ops[-1] != "(":
                apply_op()
            if not ops:
                raise FormulaError("Paréntesis desbalanceados")
            ops.pop()
        else:  # operador binario
            while ops and ops[-1] in _PRECEDENCE and _PRECEDENCE[ops[-1]] >= _PRECEDENCE[val]:
                apply_op()
            ops.append(val)

    while ops:
        if ops[-1] in "()":
            raise FormulaError("Paréntesis desbalanceados")
        apply_op()

    if len(output) != 1:
        raise FormulaError("Fórmula mal formada")
    return round(output[0], 6)
